File: backend/app/schemas/ranking.py
"""Response/request shapes for the ranking router.

Reuses the applications boundary out-models (`DimensionContributionOut`) for the
ranked candidates, so the ranked-list row and the candidate-detail contributions
share one wire shape. ``PoolDimensionOut`` is a camelCase view of the stored
``PoolDimension`` (which stays snake_case as the prompt/storage contract).
"""

from app.schemas.applications import DimensionContributionOut
from app.schemas.base import RequestModel, ResponseModel


class PoolDimensionOut(ResponseModel):
    """Camel-cased view of a stored discovery dimension."""

    key: str
    name: str
    definition: str
    high_end: str = ""
    low_end: str = ""
    why_it_differentiates: str
    from_committee_request: bool = False


class CurrentRunResponse(ResponseModel):
    """GET /ranking/current — the current run's discovered criteria, or null."""

    analysis_id: int
    dimensions: list[PoolDimensionOut]
    # The model's streamed reasoning from the discovery pass (markdown), for the
    # Insights trace. Null for runs from before it was captured / if the provider
    # surfaced none.
    discovery_narrative: str | None = None
    new_dimension_keys: list[str] = []
    # Subset of new_dimension_keys that are "revived" (seen in an earlier run, dropped,
    # now back) rather than genuinely new — derived from history at read time. The
    # frontend colours these blue ("Revived") vs. amber ("New"); new = flagged − revived.
    revived_dimension_keys: list[str] = []
    # Keys a member proposed on THIS run (from_committee_request) not yet dismissed —
    # drives the chip's "Requested" provenance pill. Cleared on the next Rank when the
    # underlying flag clears; see requested_flag_keys.
    requested_dimension_keys: list[str] = []
    kept_keys: list[str] = []
    proposed_dimensions: list[str] = []


class RawDiscoveryDimensionOut(ResponseModel):
    """One dimension as discovery emitted it, before matched keys were rewritten."""

    key: str
    name: str
    from_committee_request: bool = False


class PriorDimensionRef(ResponseModel):
    """The prior dimension a matched new dimension carried forward from: its key and
    (when known) its user-facing name. ``name`` is null for audits written before the
    prior-names capture existed."""

    key: str
    name: str | None = None


class MatchAuditResponse(ResponseModel):
    """GET /ranking/current/match-audit — the carry-forward trace for the current
    run (M13 per-run AI legibility). Null when no run exists or the run predates
    match-audit capture.

    ``carryForwardRate`` is null on a first run (no prior dimensions to match); a
    persistently near-1.0 rate on re-runs is the over-matching smell.
    """

    analysis_id: int
    raw_discovery_dimensions: list[RawDiscoveryDimensionOut]
    new_to_old: dict[str, PriorDimensionRef]  # new dimension key → the prior dim it adopted
    match_narrative: str | None = None
    prior_dimension_count: int
    discovered_count: int
    matched_count: int
    new_count: int
    carry_forward_rate: float | None = None


class SettledDimensionOut(ResponseModel):
    """One settled axis from the decomposition, for the Insights trace: what it is,
    the input axes it absorbed (``sourceKeys`` — one = kept as-is, several = a merge),
    and the model's ``decision`` reasoning (why merged / kept distinct).

    ``sourceReportMap`` maps each source key to the discovery report indices (0-based)
    that coined it — so the UI can label a source as "trade_skills (R0, R3)", showing
    which of the K discoverers surfaced it (a key in several reports = independent
    re-discovery). Empty for runs whose fan-out audit wasn't captured."""

    key: str
    name: str
    source_keys: list[str]
    source_report_map: dict[str, list[int]] = {}
    # source key → its user-facing name, so the panel labels each input axis by name +
    # key (mirroring the Matching tab). Empty for runs whose fan-out wasn't captured; the
    # UI then falls back to the bare source key.
    source_names: dict[str, str] = {}
    from_committee_request: bool = False
    decision: str


class FoldedRequestOut(ResponseModel):
    """A committee-requested axis that decomposition merged INTO another (D9): the
    request key and the settled axis it was folded into. Surfaced so a fold is visible
    to the committee, never a silent disappearance."""

    request_key: str
    into_key: str


class DecomposeAuditResponse(ResponseModel):
    """GET /ranking/current/decompose-audit — how the K fan-out discovery reports were
    settled into one non-overlapping set for the current run. Null when the run predates
    decomposition (single-discovery runs).

    ``mergeCount`` / the settle-down from ``inputDimensionCount`` to ``settledCount`` show
    how much the decomposition collapsed; ``foldedRequests`` is the D9 committee-request
    trail (empty when no request was merged away)."""

    analysis_id: int
    input_report_count: int
    input_dimension_count: int
    settled_count: int
    merge_count: int
    settled: list[SettledDimensionOut]
    folded_requests: list[FoldedRequestOut] = []
    # The decomposition pass's free-text reasoning (markdown), for the Insights panel.
    narrative: str | None = None


class ConsolidatedPairOut(ResponseModel):
    """One correlation-nominated duplicate pair and its confirm verdict, for the trace:
    ``keep``/``drop`` keys (older kept, newer aliased on a merge), their user-facing names
    (``keepName``/``dropName`` — snapshotted at consolidation time, since a merged drop key
    is removed from the report right after), the correlation ``r`` that nominated it,
    whether it ``merged``, and the model's ``reason``."""

    keep: str
    drop: str
    # Snapshotted mint names; empty when the key predates name capture (older run) — the
    # panel then falls back to the bare key.
    keep_name: str = ""
    drop_name: str = ""
    r: float
    merged: bool
    reason: str = ""


class ConsolidateAuditResponse(ResponseModel):
    """GET /ranking/current/consolidate-audit — the post-score duplicate-merge pass:
    which correlated pairs were nominated and how each was adjudicated. Null when the run
    predates the pass. ``merges`` is the applied drop→keep map; ``nominatedCount`` /
    ``mergedCount`` summarize the pass at a glance."""

    analysis_id: int
    merges: dict[str, str] = {}
    pairs: list[ConsolidatedPairOut] = []
    nominated_count: int = 0
    merged_count: int = 0
    # The confirm call's free-text reasoning (markdown), for the Insights panel.
    narrative: str | None = None


class FanOutPassOut(ResponseModel):
    """One of the K parallel discoverers, for the Insights discovery panel: the
    dimensions it found and its own reasoning narrative (null on legacy runs that stored
    reports without per-pass narratives)."""

    dimensions: list[PoolDimensionOut]
    narrative: str | None = None


class FanOutAuditResponse(ResponseModel):
    """GET /ranking/current/fan-out-audit — the K fresh-context discovery passes that
    fed decomposition, so the committee can see each discoverer (not just the one whose
    reasoning streamed live). Null on runs that predate the fan-out redesign."""

    analysis_id: int
    k: int
    passes: list[FanOutPassOut]


class RankEstimateBreakdown(ResponseModel):
    criteria_usd: float
    match_usd: float
    scoring_usd: float


class RankEstimateResponse(ResponseModel):
    """GET /ranking/run/estimate — combined cost projection for the rank chain."""

    eligible: int
    # K parallel discovery calls per Rank (the fan-out width), so the confirm card can
    # name it ("N parallel discoveries, then settle them into one set").
    fan_out: int
    breakdown: RankEstimateBreakdown
    estimated_usd: float
    approximate: bool
    cap_usd: float
    within_cap: bool
    ranking_current: bool


class ScoreCurrentEstimateResponse(ResponseModel):
    """Cost projection for filling missing scores on the current dimension set."""

    eligible: int
    to_analyze: int
    cached: int
    dimensions: int
    estimated_usd: float
    cap_usd: float
    within_cap: bool


class RankedCandidateOut(ResponseModel):
    """Camel-cased view of the ranking ``RankedCandidate`` dataclass."""

    application_id: int
    name: str | None = None
    rank: int
    fit: float
    band: str
    contributions: list[DimensionContributionOut]
    # Whether the current member has starred this applicant (private per member).
    starred_by_me: bool = False


class RankingResponse(ResponseModel):
    """GET /ranking and PUT /ranking/tiers — the ranked shortlist for the analysis."""

    # The shared analysis this ranking is for. The client echoes it back on a tier/seed
    # save so the server can reject a save against a superseded analysis (409 stale_analysis).
    analysis_id: int
    weights: dict[str, float]  # keyed by dimension key (data)
    scored_count: int
    candidates: list[RankedCandidateOut]
    new_dimension_keys: list[str] = []
    revived_dimension_keys: list[str] = []
    # Keys a member proposed on THIS run not yet dismissed — the "Requested" pill; kept
    # in sync after a tier/ack save so the badge clears in the same round-trip.
    requested_dimension_keys: list[str] = []
    kept_keys: list[str] = []
    proposed_dimensions: list[str] = []


class TierOut(ResponseModel):
    id: str
    label: str
    dimension_keys: list[str] = []
    ignore: bool = False


class TiersResponse(ResponseModel):
    """GET /ranking/tiers — the committee's importance-tier layout."""

    tiers: list[TierOut]


class SeedsResponse(ResponseModel):
    """PUT /ranking/seeds — the current pending-proposal state."""

    proposed_dimensions: list[str] = []


# --- Request bodies (camelCase-only on the wire) ----------------------------


class TierModel(RequestModel):
    id: str
    label: str
    dimension_keys: list[str] = []
    ignore: bool = False


class TierLayoutUpdate(RequestModel):
    # The analysis the client is viewing. If it isn't the current one (another member
    # re-ranked since), the save is rejected with 409 stale_analysis rather than applied
    # to the wrong board. Always current at one member, so inert until real concurrency.
    analysis_id: int
    tiers: list[TierModel]
    # Keys the committee acknowledged as "reviewed" this save (badge ✕ / "mark all
    # reviewed") — they drop out of new_dimension_keys even if left in Ignore.
    acknowledged_keys: list[str] = []
    # Keys whose "Requested" provenance pill the committee dismissed (its ✕). Separate
    # from acknowledged_keys because requested is provenance, not triage — it never
    # clears on a move, only on this explicit dismissal.
    acknowledged_requested_keys: list[str] = []


class SeedsUpdate(RequestModel):
    # The analysis the client is viewing — same stale-guard as TierLayoutUpdate.
    analysis_id: int
    # Optional so a no-op PUT leaves proposals untouched.
    proposed_dimensions: list[str] | None = None

File: frontend/src/types/ranking.ts
// The current run's discovered dimensions, from GET /ranking/current.
export type PoolDimension = {
  key: string;
  name: string;
  definition: string;
  highEnd: string;
  lowEnd: string;
  whyItDifferentiates: string;
  // True when a member proposed this axis on THIS run (per-run provenance, cleared on
  // the next Rank). Drives the chip's "Requested" pill; see enforce_committee_requests.
  fromCommitteeRequest: boolean;
};

// --- Ranking: the deterministic ranked shortlist from GET /ranking,
// pure math over the cached scores. Mirrors the backend ranking dataclasses.

// How one dimension fed a candidate's fit. `impact` = weight × (score − pool mean):
// magnitude ranks "what mattered", sign gives direction.
export type DimensionContribution = {
  dimensionKey: string;
  name: string;
  score: number;
  weight: number;
  impact: number;
  confidence: "low" | "medium" | "high";
  rationale: string;
  evidence: string;
};

export type RankedCandidate = {
  applicationId: number;
  name: string | null;
  rank: number; // 1-based position
  fit: number; // -1..+1 weighted average — supporting detail, not the headline
  band: string; // relative pool-position label (Strong fit … Limited)
  contributions: DimensionContribution[];
  // Whether the signed-in member has starred this applicant (private per member).
  starredByMe: boolean;
  // Whether the committee has placed this applicant on its shared shortlist.
  shortlisted: boolean;
};

export type RankingResponse = {
  analysisId: number;
  weights: Record<string, number>;
  scoredCount: number;
  candidates: RankedCandidate[];
  // Unacknowledged flagged dimensions (new OR revived), recomputed on every tier save
  // so badges clear in the same round-trip.
  newDimensionKeys: string[];
  // Subset of newDimensionKeys that are "revived" (seen in an earlier run, dropped,
  // now back) — the UI badges these blue vs. amber "new". new = flagged − revived.
  revivedDimensionKeys: string[];
  // Keys a member proposed on THIS run, not yet dismissed — the "Requested" pill.
  // Cleared on the next Rank when the underlying flag clears.
  requestedDimensionKeys: string[];
  // Kept axes: every dimension in a working (non-Ignore) tier — guaranteed to survive
  // the next Rank. Derived from tier placement. Echoed so the tier list stays in sync.
  keptKeys: string[];
  proposedDimensions: string[];
};

// One importance tier. Same tier → equal weight; higher tiers weigh more; Ignore
// weighs 0. The backend stores only working tiers and synthesizes the Ignore zone
// for display (the one with `ignore: true`), so the flag is optional here.
export type Tier = {
  id: string;
  label: string;
  dimensionKeys: string[];
  ignore?: boolean;
};

export type CurrentRunResponse = {
  analysisId: number;
  dimensions: PoolDimension[];
  // The model's streamed reasoning from the discovery pass (markdown), shown on the
  // Observability tab. Null when no narrative was captured.
  discoveryNarrative: string | null;
  // Flagged dimensions (new OR revived) absent from the immediately-prior run — they
  // are badged until the committee triages them. Empty on a first run.
  newDimensionKeys: string[];
  // Subset of newDimensionKeys that are "revived" (seen in an earlier run, dropped,
  // now back) — badged blue vs. amber "new". new = flagged − revived.
  revivedDimensionKeys: string[];
  // Keys a member proposed on THIS run, not yet dismissed — the "Requested" pill.
  // Cleared on the next Rank when the underlying flag clears.
  requestedDimensionKeys: string[];
  // Kept axes: every dimension in a working (non-Ignore) tier — guaranteed to survive
  // the next Rank (derived from tier placement). Plus pending free-text proposals fed
  // to the next Rank then consumed.
  keptKeys: string[];
  proposedDimensions: string[];
};

// GET /ranking/current/match-audit — the carry-forward trace for the current run.
// What discovery emitted before matched keys
// were rewritten, how the match pass mapped it onto prior dimensions, and the
// derived carry-forward rate. Null when no run or audit exists.
export type MatchAuditResponse = {
  analysisId: number;
  rawDiscoveryDimensions: { key: string; name: string; fromCommitteeRequest: boolean }[];
  // new dimension key → adopted prior dimension; name is null when unavailable.
  newToOld: Record<string, { key: string; name: string | null }>;
  matchNarrative: string | null;
  priorDimensionCount: number;
  discoveredCount: number;
  matchedCount: number;
  newCount: number;
  // Fraction matched onto a prior dimension. Null on a first run (undefined, not 0);
  // a persistently near-1.0 rate on re-runs is the over-matching smell.
  carryForwardRate: number | null;
};

// GET /ranking/current/decompose-audit — how the parallel discovery reports were
// settled into one non-overlapping dimension set. Null when no audit was recorded.
export type DecomposeAuditResponse = {
  analysisId: number;
  inputReportCount: number;
  inputDimensionCount: number;
  settledCount: number;
  mergeCount: number;
  // Each settled axis: its key/name, the input axes it absorbed (sourceKeys — one =
  // kept as-is, several = a merge), the committee-request flag, and the model's
  // decision reasoning (why merged / kept distinct).
  settled: {
    key: string;
    name: string;
    sourceKeys: string[];
    // source key → discovery report indices that coined it (e.g. {trade_skills: [0, 3]}),
    // so the UI can label a source "trade_skills (R0, R3)". Empty if fan-out uncaptured.
    sourceReportMap: Record<string, number[]>;
    // source key → its user-facing name, so a source shows as name + key (like Matching).
    // Empty if fan-out uncaptured; the UI then falls back to the bare source key.
    sourceNames: Record<string, string>;
    fromCommitteeRequest: boolean;
    decision: string;
  }[];
  // D9: committee-requested axes decomposition folded INTO another axis
  // (requestKey → intoKey), surfaced so a fold is visible, never silent.
  foldedRequests: { requestKey: string; intoKey: string }[];
  // The decomposition pass's free-text reasoning (markdown). Null if none surfaced.
  narrative: string | null;
};

// GET /ranking/current/consolidate-audit — the post-score duplicate-merge pass:
// score-vector correlation nominates suspected-duplicate pairs, a confirm call merges
// genuine ones (older key kept, newer aliased). Null on runs that predate the pass.
export type ConsolidateAuditResponse = {
  analysisId: number;
  // Applied merges: dropped (newer) key → kept (older canonical) key.
  merges: Record<string, string>;
  // Every nominated pair: keep/drop keys + their user-facing names (snapshotted at
  // consolidation time — a merged drop key leaves the report, so its name can't be
  // resolved later; empty when the key predates name capture), the correlation r that
  // flagged it, whether it merged, and the confirm call's reason.
  pairs: {
    keep: string;
    drop: string;
    keepName: string;
    dropName: string;
    r: number;
    merged: boolean;
    reason: string;
  }[];
  nominatedCount: number;
  mergedCount: number;
  // The confirm call's free-text reasoning (markdown). Null if none surfaced.
  narrative: string | null;
};

// GET /ranking/current/fan-out-audit — the parallel discoverers that fed
// decomposition. Each pass includes its dimensions and reasoning. Null when absent.
export type FanOutAuditResponse = {
  analysisId: number;
  k: number;
  passes: {
    dimensions: { key: string; name: string; definition: string; whyItDifferentiates: string }[];
    narrative: string | null;
  }[];
};
